src/instructlab/model/train.py

```diff
@@ -520,20 +580,340 @@ def get_files(directory: str, pattern: str) -> list[str]:
+
+
+def _prepare_e2e_base_dir(e2e_base_dir: str) -> None:
+    """Adds phase5 and phase10 directories and subdirectories in e2e_base_dir.
+
+    Args:
+        e2e_base_dir: top-most e2e directory, contains phase5 and phase10.
+    """
+
+    logger.debug(f"E2E -- Preparing e2e base dir: {e2e_base_dir}")
+
+    phase5_dir_path = os.path.join(e2e_base_dir, "phase5")
+    phase10_dir_path = os.path.join(e2e_base_dir, "phase10")
+
+    _create_empty_directory(phase5_dir_path)
+    _create_empty_directory(phase10_dir_path)
+
+    _setup_e2e_phase_dirs(phase5_dir_path)
+    _setup_e2e_phase_dirs(phase10_dir_path)
+
+
+def _create_empty_directory(path: str) -> None:
+    """Recursively deletes content below {path} and recreates directory."""
+    logger.debug(f"E2E -- Clearing e2e base directory: {path}")
+    if os.path.exists(path):
+        shutil.rmtree(path)
+    os.makedirs(path)
+
+
+def _setup_e2e_phase_dirs(path: str) -> None:
+    """Creates {path}/checkpoints and {path}/eval_cache directories."""
+    # TODO: these sub-directory names are hard-coded here but they
+    # could be parameterized in config.
+    logger.debug(f"E2E -- Created phase directories for {path}")
+    ckpt_path = os.path.join(path, "checkpoints")
+    eval_cache_path = os.path.join(path, "eval_cache")
+
+    os.makedirs(ckpt_path)
+    os.makedirs(eval_cache_path)
+
+
+def _e2e_training_phase(
+    train_args: TrainingArgs,
+    torch_args: TorchrunArgs,
+    data_path: str,
+    model_override: typing.Optional[str] = None,
+    ckpt_dir: typing.Optional[str] = None,
+) -> None:
+    """A single step of end-to-end training."""
+
+    from instructlab.training import (  # pylint: disable=no-name-in-module
+        run_training,
+    )
+
+    # override whatever got injected at a higher level
+    logger.debug(
+        f"E2E -- training phase -- Overriding data_path: {train_args.data_path} with {data_path}"
+    )
+    train_args.data_path = data_path
+
+    if ckpt_dir:
+        train_args.ckpt_output_dir = ckpt_dir
+
+    if model_override:
+        logger.debug(
+            f"E2E -- training phase -- Overriding model_path: {train_args.model_path} with {model_override}"
+        )
+        train_args.model_path = model_override
+
+    logger.debug(
+        f"E2E -- training phase -- Current training args for phase: {pp.pformat(train_args)}"
+    )
+
+    try:
+        run_training(train_args=train_args, torch_args=torch_args)
+    except Exception as e:
+        logging.fatal(e)
+        sys.exit(1)
+
+
+def _mmlu(ckpt_path: str) -> float:
+    from instructlab.eval.mmlu import MMLUEvaluator, MMLU_TASKS
+    import torch
+
+    evaluator = MMLUEvaluator(ckpt_path, tasks=MMLU_TASKS, few_shots=5)
+
+    ckpt_score, _ = evaluator.run()
+
+    logging.debug("E2E -- MMLU eval phase -- Clearing PyTorch cache")
+    torch.cuda.empty_cache()
+
+    return ckpt_score
+
+
+def _mtbench(ctx, ckpt_path: Path, eval_cache_path: str, judge_name: str) -> float:
```

Comment:
Is this return type correct? It says here that it's `float` but it looks like you're returning a type of `Optional[int]`

Reply:
ack